Sapling Support for z_mergetoaddress [WIP]

Eirik0 opened the pull request:
Closes #3216
Also, this is the last item in #3063.

This PR has not yet incorporated the `Sprout support to TransactionBuilder` from @str4d commit. Currently an error is thrown when trying to do mix of Sprout and Sapling notes as input, and therefore using `"ANY_ZADDR"`  or `"*"` may not work if the wallet has a mix of sprout and sapling notes.

Also, because note-locking is not yet implemented for Sapling notes the final parts of the rpc tests are do not pass.

Eirik0 commented on src/gtest/test_transaction_builder.cpp:
```diff
@@ -10,8 +10,69 @@
 static const std::string tSecretRegtest = "cND2ZvtabDbJ1gucx9GWH6XT9kgTAqfb6cotPt5Q5CyxVDhid2EN";
 
+// Fake an empty view
+class TransactionBuilderCoinsViewDB : public CCoinsView {
```
@str4d Can we not reuse
```
// Fake an empty view
class FakeCoinsViewDB : public CCoinsView {
```
from `test_validation.cpp`?

Eirik0 commented:
@zkbot try

Eirik0 commented:
I am still unable to reproduce this locally, but here is a more specific information about the failure:
```
[ RUN      ] TransactionBuilder.Invoke
unknown file: Failure
C++ exception with description "Could not create joinsplit description" thrown in the test body.
[  FAILED  ] TransactionBuilder.Invoke (64532 ms)
```

Eirik0 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4458,7 +4453,7 @@ UniValue z_mergetoaddress(const UniValue& params, bool fHelp)
         Params().GetConsensus(),
         nextBlockHeight);
-    bool isShielded = numNotes > 0 || isToZaddr;
+    bool isShielded = numNotes > 0 || isToSproutZaddr;
```
This ends up getting confused in the next commit as it should only be looking at Sprout notes. This is not a problem per se because the `contextualTx` below is only used if there are no sapling notes (currently). I am rebasing and fixing this to make it more clear what is going on.

Eirik0 commented:
Force pushed some minor changes to address my above comment.

Eirik0 commented:
Removed the changes to TransactionBuilder (Some of which I split out into https://github.com/zcash/zcash/pull/3646) and force pushed.

Eirik0 commented:
@zkbot try

Eirik0 commented:
One thing to mention is that the assertions in `rpc_z_mergetoaddress_parameters` were not doing what they said they were doing as we had not ever set the experimental flags to allow using z_mergetoaddress, for example:
```
// bad from address
BOOST_CHECK_THROW(CallRPC("z_mergetoaddress "
"** tnpoQJVnYBZZqkFadj2bJJLThNCxbADGB5gSGeYTAGGrT5tejsxY9Zc1BtY8nnHmZkB"), runtime_error);
```
correctly passed the `BOOST_CHECK_THROW`, but for the wrong reason. The runtime_error being thrown was: `JSONRPCError(RPC_WALLET_ERROR, "Error: z_mergetoaddress is disabled.");` wrapped in a runtime_error (`CallRPC` catches JSONRPCErrors and rethrows them as runtime_errors). 

That being said, my assessment was that the underlying code in z_mergetoaddress was not incorrect. In any event, this is now fixed.

LarryRuane commented on src/wallet/asyncrpcoperation_mergetoaddress.h:
```diff
@@ -25,10 +26,12 @@
 using namespace libzcash;
 
 // Input UTXO is a tuple of txid, vout, amount
```
update comment (for Cscript)?

Eirik0 replied:
Whoops... missed that. I was tempted to just delete some of these comments because I think that these tuples are self-explanatory. That being said, I'll update it to add `, script`.

Eirik0 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4254,21 +4251,26 @@ UniValue z_mergetoaddress(const UniValue& params, bool fHelp)
+                isToSproutZaddr = true;
             }
         } else {
             throw JSONRPCError(RPC_INVALID_PARAMETER, string("Invalid parameter, unknown address format: ") + destaddress );
```
I am not a huge fan of the above block of code ~(4262 - 4280) and went back and forth a lot when working on it. I had considered at one point adding a methods `IsSproutPaymentAddress` and `isSaplingPaymentAddress` akin to `IsValidPaymentAddress` which I think would could make this more readable. I also thing it's confusing that we have the mutually exclusive bools `isToSproutZaddr` and `isToSaplingZaddr` and considered what this would look like if there were an enum in their stead. That being said, I eventually decided to move on. For this reason, if it is essential, I think I would be happier to deal with cleaning this up in a separate PR.

Eirik0 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4254,21 +4251,26 @@ UniValue z_mergetoaddress(const UniValue& params, bool fHelp)
+                isToSproutZaddr = true;
             }
         } else {
             throw JSONRPCError(RPC_INVALID_PARAMETER, string("Invalid parameter, unknown address format: ") + destaddress );
```
This comment appears to be a duplicate.

bitcartel commented:
@Eirik0 Maybe change text "send from .. to" to be "send between Sprout and Sapling"?
```
./zcash-cli z_mergetoaddress '["ztestsapling1ehdd46fqkx39762r3vagzp30cv9lyscphnz8qz3lgw7xyq5xufx96nhvg5klh0wcuxhp7nuw5l5"]' ztVjJST6iAbPKXLpyvLbyUqXL93aqPn6ek3fjVgytQGj76BMiuTTuTVZvdr5tNEHhH3it92w6htki9rG4agBbRdkFtDRdog
error code: -8
error message:
Cannot send from Sprout to Sapling addresses using z_mergetoaddress
```

Eirik0 commented:
@bitcartel
Would you suggest changing both sprout->sapling and sapling->sprout to be:
`"Cannot send between Sprout and Sapling addresses using z_mergetoaddress"`
or having the above and `"... Sapling and Sprout ..."` for the sapling->sprout case?

bitcartel commented:
@Eirik0 I think the former works and could lead to simpler code as one less condition to check.

Eirik0 commented:
@bitcartel  On inspection, I noticed that we do pretty much the same thing for z_sendmany, so I am going to change both z_mergetoaddress' error message and that.

Eirik0 commented:
@bitcartel
> The sprout qa test takes a long time, could it be shortened?

My changed were intended to be a straight refactor, so this test should function exactly the same as before. It probably could be shortened, but I found the coverage it provides helpful when adding the Sapling support. Also I think there is advantage to having both the Sprout and Sapling tests do the same thing (which they do currently).

> Would be nice if the commits could be squashed into logical chunks of work.

Agreed. I will do this after having addressed the review comments.

Eirik0 commented:
@bitcartel I've addressed your comments. I will look in to squashing things together once you've double checked those.

Eirik0 commented:
Reorganized and squashed some of the the commits.

Before rebasing/plan:
```
3a. e464a3d8f Rename mergetoaddress rpc test
3b. 340cee0cf Move test logic to helper class
3c. 9a7a5e32a cleanup and format
3d. 21692f684 Add mergetoaddress test for sapling
2. 76192e703 Add fail method to rpc test utils
3e. 6b0962dea Show more information when getting an unexpected exception
4a. a0233441f Remove pre-Sapling error check
4b. 0fea484d2 Reorganize checking for Sprout and Sapling
4c. b6129eabb Add sapling support to z_mergetoaddress
4d. c8e7c1f14 Sprout z_mergetoaddress inputs use Sprout spending keys
5. 4c15ac1a3 Add locking for Sapling notes
8. 7bf420a80 Adjust z_mergetoaddress assertions in Sapling rpc test
4e. 812af106a Replace ANY_ZADDR with ANY_SPROUT or ANY_SAPLING and remove *
1. 13b4f06c3 Fix z_mergetoaddress parameter tests
7. 9374b04d0 Add additional z_mergetoaddress parameter tests
4f. 5266dbf8a Increate default note limit for z_mergetoaddress
4g. 5218cf6cc fix comments and other minor cleanup
6. e5fbe69b8 Better error messages when sending between Sprout and Sapling
4i. 3cbfa3135 Do not allow mergetoaddress operations to be constructed with both Sprout and Sapling inputs
```
After rebasing:
```
1. Fix z_mergetoaddress parameter tests
2. Add fail method to rpc test utils
3. Extentend sprout mergetoaddress rpc test to also work for sapling
4. Add Sapling support to z_mergetoaddress
5. Add locking for Sapling notes
6. Better error messages when sending between Sprout and Sapling
7. Add additional z_mergetoaddress parameter tests
8. Adjust z_mergetoaddress assertions in Sapling rpc test
```
As a note, I didn't run in to any issues while doing this, the lines added/removed still match, and the `mergetoaddress_sprout.py` and `mergetoaddress_sapling.py` tests both pass for me locally.

bitcartel commented:
@Eirik0 Looks good, nice squashing!

Non-blocking, there is spelling mistake in commit message for 3: 'Extentend' 

Otherwise, rebase (if necessary) after #3680 has merged and we can merge this PR.

Eirik0 commented:
Lol, whoops! Looks like I used the extended spelling :)

bitcartel commented:
@zkbot r+
